gctscript/modules/ta/indicators/ema.go

```diff
@@ -0,0 +1,56 @@
+package indicators
+
+import (
+	"fmt"
+
+	objects "github.com/d5/tengo/v2"
+	"github.com/thrasher-corp/go-talib/indicators"
+	"github.com/thrasher-corp/gocryptotrader/gctscript/modules"
+	"github.com/thrasher-corp/gocryptotrader/gctscript/wrappers/validator"
+)
+
+var EMAModule = map[string]objects.Object{
+	"calculate": &objects.UserFunction{Name: "calculate", Value: ema},
+}
+
+func ema(args ...objects.Object) (objects.Object, error) {
+	if len(args) != 2 {
+		return nil, objects.ErrWrongNumArguments
+	}
+
+	ohlcvInput := objects.ToInterface(args[0])
+	ohlcvInputData, valid := ohlcvInput.([]interface{})
+	if !valid {
+		return nil, fmt.Errorf(modules.ErrParameterConvertFailed, OHLCV)
+	}
+
+	var ohlcvClose []float64
+	for x := range ohlcvInputData {
+		switch t := ohlcvInputData[x].(type) {
+		case []interface{}:
+			value, err := toFloat64(t[4])
+			if err != nil {
+				return nil, err
+			}
+			ohlcvClose = append(ohlcvClose, value)
+		default:
+			return nil, fmt.Errorf(modules.ErrParameterConvertFailed, OHLCV)
+		}
+	}
+
+	inTimePeriod, ok := objects.ToInt(args[1])
+	if !ok {
+		return nil, fmt.Errorf(modules.ErrParameterConvertFailed, inTimePeriod)
+	}
+
+	var ret []float64
+	if validator.IsTestExecution.Load() != true {
```

Comment:
not happy with this being in production code going to make a quick change now to convert this to the wrapper/module setup we use for the rest of engine linkup and hook up to validator correctly

Reply:
Sounds good!

Reply:
Really small nit, I think these could be changed around across indicators to reduce nesting of the expected path and exit sooner
```
	r := &objects.Array{}
	if validator.IsTestExecution.Load() == true {
		return r, nil
	}
	ret = indicators.Ma(ohlcvClose, inTimePeriod, indicators.EMA)
```

Reply:
yeah will make this execution flow change now

Reply:
Also will move it much further up as none of the additional steps are required part of a validator execution